util/tracing/tracer_span.go

```diff
@@ -0,0 +1,568 @@
+func IsNoopContext(spanCtx opentracing.SpanContext) bool {
+	_, noop := spanCtx.(noopSpanContext)
+	return noop
+}
+
+// SetSpanStats sets the stats on a span. stats.Stats() will also be added to
+// the span tags.
+func SetSpanStats(os opentracing.Span, stats SpanStats) {
+	s := os.(*span)
+	s.mu.Lock()
+	s.mu.stats = stats
+	for name, value := range stats.Stats() {
+		s.setTagInner(StatTagPrefix+name, value, true /* locked */)
+	}
+	s.mu.Unlock()
+}
+
+// Finish is part of the opentracing.Span interface.
+func (s *span) Finish() {
+	s.FinishWithOptions(opentracing.FinishOptions{})
+}
+
+// FinishWithOptions is part of the opentracing.Span interface.
+func (s *span) FinishWithOptions(opts opentracing.FinishOptions) {
+	finishTime := opts.FinishTime
+	if finishTime.IsZero() {
+		finishTime = time.Now()
+	}
+	s.mu.Lock()
+	s.mu.duration = finishTime.Sub(s.startTime)
+	s.mu.Unlock()
+	if s.shadowTr != nil {
+		s.shadowSpan.Finish()
+	}
+	if s.netTr != nil {
+		s.netTr.Finish()
+	}
+}
+
+// Context is part of the opentracing.Span interface.
+func (s *span) Context() opentracing.SpanContext {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	baggageCopy := make(map[string]string, len(s.mu.Baggage))
+	for k, v := range s.mu.Baggage {
+		baggageCopy[k] = v
+	}
+	sc := &spanContext{
+		spanMeta: s.spanMeta,
+		Baggage:  baggageCopy,
+	}
+	if s.shadowTr != nil {
+		sc.shadowTr = s.shadowTr
+		sc.shadowCtx = s.shadowSpan.Context()
+	}
+
+	if s.isRecording() {
+		sc.recordingGroup = s.mu.recordingGroup
+		sc.recordingType = s.mu.recordingType
+	}
+	return sc
+}
+
+// SetOperationName is part of the opentracing.Span interface.
+func (s *span) SetOperationName(operationName string) opentracing.Span {
+	if s.shadowTr != nil {
+		s.shadowSpan.SetOperationName(operationName)
+	}
+	s.operation = operationName
+	return s
+}
+
+// SetTag is part of the opentracing.Span interface.
+func (s *span) SetTag(key string, value interface{}) opentracing.Span {
+	return s.setTagInner(key, value, false /* locked */)
+}
+
+func (s *span) setTagInner(key string, value interface{}, locked bool) opentracing.Span {
+	if s.shadowTr != nil {
+		s.shadowSpan.SetTag(key, value)
+	}
+	if s.netTr != nil {
+		s.netTr.LazyPrintf("%s:%v", key, value)
+	}
+	if s.isRecording() {
+		if !locked {
+			s.mu.Lock()
+		}
+		if s.mu.tags == nil {
+			s.mu.tags = make(opentracing.Tags)
+		}
+		s.mu.tags[key] = value
+		if !locked {
+			s.mu.Unlock()
+		}
+	}
+	return s
+}
+
+// LogFields is part of the opentracing.Span interface.
+func (s *span) LogFields(fields ...otlog.Field) {
+	if s.shadowTr != nil {
+		s.shadowSpan.LogFields(fields...)
+	}
+	if s.netTr != nil {
+		// lights step does not support arbitrary event
+		if len(fields) == 1 && fields[0].Key() == "event" {
+			s.netTr.LazyPrintf("%s", fields[0].Value())
+		} else {
+			var buf bytes.Buffer
+			for i, f := range fields {
+				if i > 0 {
+					buf.WriteByte(' ')
+				}
+				fmt.Fprintf(&buf, "%s:%v", f.Key(), f.Value())
+			}
```

Comment:
how about constract format string  ``%s%v %s%v %s%v....`` and args `[]interface{....}`

and here we give up lazy feature, I think not lazy is better too, because we don't know how bug data user pass in :laughing: